### PR TITLE
 Add ML-based Pod/Container Failure Detection Module (Synthetic Data)

### PR DESCRIPTION
**Pull request by radheshpai87**
### Overview
Introduces a standalone `ml-models/` module that trains and serves three machine learning classifiers to detect impending pod and container failures in Docker and Kubernetes environments. The module is fully self-contained with Docker support, a REST inference API, and automated model comparison reporting.

---

#### Dataset
- Added `pod_failure_dataset.csv` (50,000 labelled pod observations) to `ml-models/datasets/`
- Features: `cpu_usage`, `memory_usage`, `disk_usage`, `network_latency`, `restart_count`, `probe_failures`, `node_cpu_pressure`, `node_memory_pressure`, `pod_age_minutes`
- Target: `failure_label` (0 = healthy, 1 = failure)

#### Models (`ml-models/models/`)
Three independent binary classifiers, each with a `train.py` and `predict.py`:

| Model | Framework | Notes |
|---|---|---|
| Random Forest | scikit-learn | 200 estimators, balanced class weights |
| XGBoost | XGBoost | 300 estimators, scale_pos_weight for imbalance |
| LSTM | PyTorch | Stacked 2-layer LSTM, BCEWithLogitsLoss, GPU-ready |

#### Inference API (`ml-models/api/`)
- FastAPI server exposing 5 endpoints:
  - `GET /health` — lists loaded models
  - `POST /predict/random-forest`
  - `POST /predict/xgboost`
  - `POST /predict/lstm`
  - `POST /predict/ensemble` — majority-vote across all loaded models
- Pydantic v2 request/response schemas with validation
- Models loaded lazily on startup; missing artifacts are skipped gracefully

#### Model Comparison (`ml-models/compare_models.py`)
Automatically runs after `train_all.py` and outputs to `ml-models/reports/`:
- `metrics_comparison.png` — grouped bar chart (Accuracy / F1 / ROC-AUC)
- `roc_curves.png` — overlaid ROC curves
- `confusion_matrices.png` — 1×3 heatmap grid
- `metrics_summary.txt` — plaintext table with best model recommendation

#### Docker Setup
- Multi-stage `Dockerfile` with `train` and `api` targets
- `docker-compose.yml` with separate `train` (one-shot) and `api` services
- GPU support via `TORCH_INDEX_URL` build arg — switch from CPU to CUDA 11.8/12.1 with a single line change
- Model artifacts persisted to host via bind-mount volumes

---

### Results (from `metrics_summary.txt`)

```
Model               Accuracy  Precision   Recall       F1   ROC-AUC
Random Forest         0.9913     0.7530   0.7310   0.7418    0.9963 ◄ BEST
XGBoost               0.9884     0.6190   0.8363   0.7114    0.9957
LSTM                  0.9462     0.2390   0.9825   0.3844    0.9901
```

**Recommended model: Random Forest** — highest ROC-AUC (0.9963) and best precision/F1 balance. Key finding: `cpu_usage` and `memory_usage` simultaneously above ~90% are the dominant failure predictors.

---

### Usage

```bash
# Train all models + generate comparison charts
docker compose run --rm train

# Start inference API
docker compose up api

# Swagger UI
open http://localhost:8000/docs
```

---

### Notes
- `ml-models/` is intentionally kept separate from `backend/` — integration can be done in a follow-up PR
- The `train` service is decoupled from `api` (`depends_on` removed) to prevent race conditions during startup
- LSTM precision is low (0.24) due to the model being recall-optimised; for production use the Random Forest or ensemble endpoint

### CODE REVIEW

**Review by abbysallord — APPROVED** [review]
Looks good to me. I will merge it
